x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,212 @@
+	a.EventsToBeTracked = atomic.NewUint64(publishedEvent)
+	a.ReceiveCount = receiveCount
+	a.start = start
+	a.processingErr = processingErr
+	a.Handles = handles
+	a.keepaliveCancel = keepaliveCancel
+	a.keepaliveWg = keepaliveWg
+	a.msgHandler = msgHandler
+	a.log = log
+}
+
+func (a *EventACKTracker) FullyTracked() bool {
+	eventsToBeTracked := a.EventsToBeTracked.Load()
+	if eventsToBeTracked == 0 {
+		return false
+	}
+
+	// This is eating its own tail: we should check for dropped+published, but then we won't wait for acked.
+	// Acked might not be equal to published?
+	return a.EventsDropped.Load()+a.EventsAcked.Load() == eventsToBeTracked
+}
+
+// FlushForSQS delete related SQS message
+func (a *EventACKTracker) FlushForSQS() {
+	// Stop keepalive visibility routine before deleting.
+	a.keepaliveCancel()
+	a.keepaliveWg.Wait()
+
+	eventsPublished := a.EventsPublished.Load()
+	if eventsPublished > 0 && eventsPublished == a.EventsAcked.Load() {
+		err := a.msgHandler.DeleteSQS(a.msg, a.ReceiveCount, a.processingErr, a.Handles)
+		if err != nil {
+			a.log.Warnw("Failed deleting SQS message.",
+				"error", err,
+				"message_id", *a.msg.MessageId,
+				"elapsed_time_ns", time.Since(a.start))
+		} else {
+			a.log.Debugw("Success deleting SQS message.",
+				"message_id", *a.msg.MessageId,
+				"elapsed_time_ns", time.Since(a.start))
+		}
+	} else {
+		a.log.Infow("Skipping deleting SQS message, not all events acked.",
+			"events_acked", a.EventsAcked.Load(),
+			"events_dropped", a.EventsDropped.Load(),
+			"events_published", a.EventsPublished.Load(),
+			"events_tracked", a.EventsToBeTracked.Load(),
+			"message_id", *a.msg.MessageId,
+			"elapsed_time_ns", time.Since(a.start))
+
+	}
+
+	a.deletionWg.Done()
+
+}
+
+// ACK increments the number of EventsAcked.
+func (a *EventACKTracker) ACK() {
+	a.EventsAcked.Inc()
+}
+
+// Drop increments the number of EventsDropped.
+func (a *EventACKTracker) Drop() {
+	a.EventsDropped.Inc()
+}
+
+// Published increments the number of EventsPublished.
+func (a *EventACKTracker) Published() {
+	a.EventsPublished.Inc()
+}
+
+// NewEventACKHandler returns a beat ACKer that can receive callbacks when
+// an event has been ACKed an output. If the event contains a private metadata
+// pointing to an eventACKTracker then it will invoke the trackers ACK() method
+// to decrement the number of pending ACKs.
+func NewEventACKHandler() beat.EventListener {
+	return acker.ConnectionOnly(
+		acker.Combine(
+			newEventListener(),
+			acker.EventPrivateReporter(func(_ int, privates []interface{}) {
+				for _, private := range privates {
+					if acker, ok := private.(*EventACKTracker); ok {
+						acker.ACK()
+					}
+				}
+			}),
+		))
+}
+
+func newEventListener() *eventListener {
+	return &eventListener{}
+}
+
+type eventListener struct{}
+
+func (a *eventListener) ACKEvents(n int) {}
+
+func (a *eventListener) ClientClosed() {}
+
+func (a *eventListener) AddEvent(event beat.Event, published bool) {
```

Comment:
`EventPrivateReporter` already triggers your callback in `NewEventACKHandler` for every event, whether it is published or not, so the current `Drop` handling leads to double counting. I believe you can remove `eventListener` entirely, along with `EventsDropped` and `EventsPublished` -- there should be no reason to alter behavior based on whether something is published or dropped, since you just need to compare the final event count for a message with the callback invocations from `EventPrivateReporter`.

Reply:
I can see that in `AddEvent` I have the `published` param, that tells me if the event was published or dropped.

> EventPrivateReporter already triggers your callback in NewEventACKHandler for every event, whether it is published or not

Oki, I thought it trigger my callback for every "acked" event (as " output returned a positive response of receiving the event").

If "acked" means that the queue handled the message, either publishing it or dropping the, yes. Then I can just use only `EventPrivateReporter` and `LastEventPrivateReporter`.

In `EventPrivateReporter` I will just call `acker.ACK()` for every event, and in `LastEventPrivateReporter`:
```
if acker.FullyTracked() {
    acker.FlushForSQS()
}
```

It seems this way we don't have the possibility to ensure at-least-once delivery: at least in the meaning that the output ingested the event.